setup xtask infra for cheatsheets

miguelraz opened the pull request:
## Problem:

 Making cheatsheets like "From Python to Rust" and friends is going to be a maintainability nightmare as `training-materials` grows, changes ownership, etc., over time. This was raised as a concern in #148 and this PR seeks to fix that.

## Proposed solution:

Setup a `xtask` style workflow where we 

* call `cargo xtask make-cheatsheet python` at the root folder
* scrape Markdown headers in `SUMMARY..md` and segment topics by `Rust Fundamentals`, `Applied Rust` and `Advanced Rust`
* write out to `src/python-cheatsheet.md` if it doesn't exist
* if it does exist, check that it in sync: all headers in `python-cheatsheet.md` are in the appropriate sections, in order, and none are missing.

The code is (I think) defensive to spare future pain for someone else looking at this code ~~(myself in a few months)~~.

This PR only sets up the infrastructure but doesn't add any cheatsheets themselves - I'll file a separate PR with a basic Julia cheatsheet once this lands to separate concerns.

-------

<details>
<summary> Current functionality: </summary>

```text
❯ cargo xtask make-cheatsheet python
   Compiling xtask v0.1.0 (C:\Users\mrg\work\rust-training\xtask)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.10s
     Running `target\debug\xtask.exe make-cheatsheet python`
Cheatsheat for python written at ./training-slides/src/python-cheatsheet.md

❯ cat .\training-slides\src\python-cheatsheet.md
# Rust Fundamentals
## Overview
## Installation
## Basic Types
## Control Flow
## Compound Types
## Ownership and Borrowing
## Error Handling
## Collections
## Iterators
## Imports and Modules
## Good Design Practices

# Applied Rust
## Methods and Traits
## Rust I/O Traits
## Generics
## Lifetimes
## Cargo Workspaces
## Heap Allocation (Box and Rc)
## Shared Mutability (Cell, RefCell)
## Thread Safety (Send/Sync, Arc, Mutex)
## Closures and the Fn/FnOnce/FnMut traits
## Spawning Threads and Scoped Threads

# Advanced Rust
## Advanced Strings
## Debugging Rust
## Dependency Management with Cargo
## Deref Coercions
## Design Patterns
## Documentation
## Drop, Panic and Abort
## Dynamic Dispatch
## Macros
## Property Testing
## Rust Projects Build Time
## Send and Sync
## Serde
## Testing
## The stdlib
## Using Cargo
## Using Types to encode State
## WASM

❯ cargo xtask make-cheatsheet python
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `target\debug\xtask.exe make-cheatsheet python`
File python-cheatsheet.md already exists - checking it's in sync
Neat! python-cheatsheet.md is in sync

rust-training on  xtask-cheatsheet [$?] via 🦀 v1.78.0
❯ cargo xtask test-cheatsheet python
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `target\debug\xtask.exe test-cheatsheet python`
Neat! python-cheatsheet.md is in sync
```

</details>

jonathanpallant commented:
Could we add a few words (here, or in a README) about how the cheatsheets work - to save having to read and understand the source code?

jonathanpallant commented:
Also, it needs a rebase on `main` to fix the CloudFlare CI failure, and some thought about the workspaces to fix the GHA CI failure.

miguelraz commented:
@jonathanpallant 

* added a README.md
* rebased to main to have Cloudflare deploy passing
* fixed workspace issue by adding a `exclude = ["examples/"]` array to the `Cargo.toml` and CI is happy

jonathanpallant commented on xtask/README.md:
```diff
@@ -0,0 +1,79 @@
+
+## How
+
+As `training-material` grows and changes, maintenance could be a nightmare. We basically don't want to be in the business of remember that a certain slide got reordered or moved from one section to another and thus needs changing in all the cheathseets as well. Therefore, this tool seeks to alleviate that with the following workflow:
```
```suggestion
As `training-material` grows and changes, maintenance could be a nightmare. We basically don't want to be in the business of remembering that a certain slide got reordered or moved from one section to another and thus needs changing in all the cheatsheets as well. Therefore, this tool seeks to alleviate that with the following workflow:
```

jonathanpallant commented on xtask/README.md:
```diff
@@ -0,0 +1,79 @@
+As `training-material` grows and changes, maintenance could be a nightmare. We basically don't want to be in the business of remember that a certain slide got reordered or moved from one section to another and thus needs changing in all the cheathseets as well. Therefore, this tool seeks to alleviate that with the following workflow:
+
+* call `cargo xtask make-cheatsheet python` at the root folder
+* scrape Markdown headers in `SUMMARY.md` and segment topics by `Rust Fundamentals`, `Applied Rust` and `Advanced Rust`
```
Who does this - the user or the tool? Because step 1 in this list is something the user has to do.

miguelraz replied:
Yes, for now it's the user's responsibility to run the check after modifying any of the cheatsheets - I wouldn't mind doing a follow up PR that adds it to the build scripts. I'll add that as a follow up task for myself on Clickup.

jonathanpallant commented on xtask/src/main.rs:
```diff
@@ -0,0 +1,43 @@
+
+    // first arg is the name of the executable; skip it
+    let args = env::args().skip(1).collect::<Vec<_>>();
+    let args = args.iter().map(|arg| &arg[..]).collect::<Vec<_>>();
```
What is this line for? `args` is already a `Vec<String>` right?

jonathanpallant commented on xtask/src/main.rs:
```diff
@@ -0,0 +1,43 @@
+    let langs = vec!["python", "go", "cpp", "swift", "java", "julia", "c"];
+
+    if !langs.contains(&args[1]) {
+        panic!("{} is not a valid language name. \nExpected one of: python, go, cpp, swift, java, julia, c", &args[1]);
```
I'd print the lang list in the panic so it doesn't get out of sync.

miguelraz replied:
Good call, added that and some cleaner formatting in the print.

jonathanpallant commented on xtask/src/tasks.rs:
```diff
@@ -0,0 +1,260 @@
+}
+
+#[test]
+fn test_get_slide_name() {
```
Tests should go in a `#[cfg(test)] mod test { ... }`

miguelraz replied:
Neater, done.

jonathanpallant commented on xtask/src/tasks.rs:
```diff
@@ -0,0 +1,260 @@
+        panic!("YOUR LAST_HEADER is not part of the text input. CHECK your `SUMMARY.md` for {LAST_HEADER}");
+    }
+
+    let first_header = text.find(INITIAL_HEADER).unwrap();
```
You can handle the error here rather than having a separate `contains` check which might get isolated from this line.

```rust
let Some(first_header) = text.find(INITIAL_HEADER) else {
    panic!("Could not find the initial header {INITIAL_HEADER:?}. Check your `SUMMARY.md`");
}
```

miguelraz replied:
Oh, that is a much nicer form. Done.

jonathanpallant commented on xtask/src/tasks.rs:
```diff
@@ -0,0 +1,260 @@
+}
+
+#[test]
+fn test_focus_regions() {
```
Tests should go in a #[cfg(test)] mod test { ... }

miguelraz replied:
Also refactored, thanks.

jonathanpallant commented on xtask/src/tasks.rs:
```diff
@@ -0,0 +1,260 @@
+}
+
+#[test]
+fn test_extract_slides() {
```
Tests should go in a #[cfg(test)] mod test { ... }

miguelraz replied:
Also refactored.

jonathanpallant commented:
OK, I think this tool makes sense. I had some minor points about the program itself, but I also think we should be running the unit tests in CI, and testing our cheat-sheets in CI.

miguelraz commented on xtask/src/main.rs:
```diff
@@ -0,0 +1,43 @@
+    let args = env::args().skip(1).collect::<Vec<_>>();
+    let args = args.iter().map(|arg| &arg[..]).collect::<Vec<_>>();
+
+    let langs = vec!["python", "go", "cpp", "swift", "java", "julia", "c"];
```
Yup!

miguelraz replied:
Done.

jonathanpallant commented:
re-assign to me when you're ready, thanks!

miguelraz commented on xtask/src/main.rs:
```diff
@@ -0,0 +1,43 @@
+        ["test-cheatsheet", lang] => tasks::test_cheatsheet(lang),
+        _ => {
+            eprintln!(
+                "cargo xtask
```
Good call, done.

miguelraz commented:
Force pushed to get rid of a pesky rebase muddying the history horribly.

miguelraz commented:
Should be ready for review again.

jonathanpallant commented on xtask/src/main.rs:
```diff
@@ -0,0 +1,45 @@
+    test-cheatsheet [LANG]      test LANG's cheatsheet (all `SUMMARY.md` items are in sheet and viceversa)
+
+LANG:
+    Valid values are `python, go, cpp, swift, java, julia, c`
```
I would replace this list with a magic token like `$$LANG_LIST$$` and do `HELP_TEXT.replace("$$LANG_LIST$$", langs.join(", ")` this string before you print it, like:

```rust
static HELP_TEXT: &'static str = "cargo xtask

LANG:

  We support $$LANG_LIST$$
";

fn main() {
    let lang_list = ["Apples", "Bananas", "Pears"];
    println!("{}", HELP_TEXT.replace("$$LANG_LIST$$", &join_str(&lang_list)));
}

fn join_str(input: &[&str]) -> String {
    let mut output = String::new();
    let mut iter = input.iter().peekable();
    while let Some(lang) = iter.next() {
        if iter.peek().is_none() {
            output += " and ";
            output += lang;
        } else {
            output += ", ";
            output += lang;
        }
    }
    output
}
```

jonathanpallant replied:
the fancy join is optional

miguelraz replied:
Fancy indeed. I'll work that in.

jonathanpallant commented:
There's quite a lot of code here, and coming to it fresh after vacation I lacked any real context. Some file-level comments explaining the purpose of the file and some function-level comments explaining the purpose of each function would be a useful addition.

miguelraz commented:
I'll write those in later today, good point.

miguelraz commented:
@jonathanpallant this feels ready for review again.

jonathanpallant commented:
I elected to push some changes rather than leave a bunch of review comments. Please do review the diff and let me know if you have any questions.

The final matter is deciding how we want CI to run this, to automatically check out cheatsheets. Ideally we'd have a `cargo xtask test-cheatsheets` which automatically tested all of them.

miguelraz commented:
All the comments being hoisted to a README make sense to me and the added level of doc comments is welcome, thanks!

I'm in favor of writing a `cargo xtask test-all`.

miguelraz commented:
Resolved a small merge conflict by accepting the `xtask` `target` folder as part of `.gitignore`

miguelraz commented:
Ah, I noticed there's a corner case I haven't handled, I'm going to work on it for a bit.
